File: Super_Egg_drop_DP.py
# DP Solution
# Time complexity : O(n*n*k)
# Space complexity : O(n*k)
# Leetcode : Time limit exceeded

class Solution:
    def superEggDrop(self, k: int, n: int) -> int:
        # make a dp array of size n+1 * k+1
        dp = [[0 for _ in range(n+1)] for _ in range(k+1)]
        
        # filling first row
        # fill the row with the value of the floors
        for i in range(1,n+1):
            dp[1][i] = i
        
        # rows
        # traversing through each row starting from 2nd row
        for i in range(2,k+1):
            # cols
            # traverse through all cols starting with 1st
            for j in range(1,n+1):
                # floor
                # filling the values with Infinity as we need to find the min
                dp[i][j] = float('inf')
                
                # go over each floor which ranges from 0 till j+1, as my current floors
                # this is the loop for permutations
                # break - egg-1 and remaining floors below the curr floor
                # no break - same eggs (k) and remaining floors above the curr floor
                for f in range(j+1):
                    # find the max between the break and no break case and add one to it
                    dp[i][j] = min(dp[i][j], 1 + max(dp[i][j-f], dp[i-1][f-1]))
        
        # return the final value calculated in the last index
        return dp[k][n]

File: SuperEggDrop/SuperEggDrop_DP.java
/* Time Complexity : O((n^2)*k) 
 *  n - no. of floors */
/* Space Complexity : O(n*k)
 * k - no. of eggs */
// Did this code successfully run on Leetcode : Yes 
// Any problem you faced while coding this : TLE


class Solution {
    public int superEggDrop(int k, int n) {
        int[][] dp = new int[k+1][n+1];
        //Populating row with one egg
        for(int j = 1; j <= n; j++){
            dp[1][j] = j;
        }
        //Populating col with one floor
        for(int i = 1; i <= k; i++){
            dp[i][1] = 1;
        }

        //i: range for eggs
        for(int i = 2; i <= k; i++){
            //i: range for floors
            for(int j = 2; j <= n; j++){
                int min = Integer.MAX_VALUE;
                for(int f = 1; f <= j; f++){
                    // curr no. of moves = 1 + Max(break, not broken)
                    //break = dp[eggs left][row/floors below curr row/floor]
                    //not broken = dp[total eggs][row/floors above curr row/floor]
                    int currMoves = 1 + Math.max(dp[i-1][f-1], dp[i][j-f]);
                    min = Math.min(min, currMoves);
                }
                dp[i][j] = min;
            }
        }
        return dp[k][n];
    }
}